src/parsemultipart.jl

```diff
@@ -0,0 +1,269 @@
+    # the end of the boundary delimiter, and the \r\n.
+    # The last boundary delimiter line has an additional '--' at the end of the boundary delimiter
+    # [RFC2046 5.1.1](https://tools.ietf.org/html/rfc2046#section-5.1.1)
+
+    i = start
+    end_index = i+length(boundaryDelimiter)-1
+    while end_index <= length(bytes)
+        if bytes[i:end_index] == boundaryDelimiter
+            # boundary delimiter line start on a new line ...
+            if i > 1
+                (i == 2 || bytes[i-2] != CR_BYTE || bytes[i-1] != LF_BYTE) && error("boundary delimiter found, but it was not the start of a line")
+                # the CRLF preceding the boundary delimiter is "conceptually attached
+                # to the boundary", so account for this with the index
+                i-=2
+            end
+
+            # need to check if there are enough characters for the CRLF or for two dashes
+            end_index < length(bytes)-1 || error("boundary delimiter found, but did not end with new line")
+
+            is_terminating_delimiter = bytes[end_index+1] == DASH_BYTE && bytes[end_index+2] == DASH_BYTE
+            is_terminating_delimiter && (end_index+=2)
+
+            # ... there can be arbitrary SP and HTAB space between the boundary delimiter ...
+            while (end_index < length(bytes) && (bytes[end_index+1] in [HTAB_BYTE, SPACE_BYTE])) end_index+=1 end
+            # ... and ends with a new line
+            (end_index < length(bytes)-1) && bytes[end_index+1] == CR_BYTE && bytes[end_index+2] == LF_BYTE || error("boundary delimiter found, but did not end with new line")
+            end_index += 2
+
+            return (is_terminating_delimiter, i, end_index)
+        end
+
+        i += 1
+        end_index += 1
+    end
+
+    error("boundary delimiter not found")
+end
+
+"""
+    find_multipart_boundaries(bytes, boundary; start=1)
+
+Find the start and end indexes of all the parts of the multipart object.  Ultimately this method is
+looking for the data between the boundary delimiters in the byte array.  A vector containing all
+the start/end pairs is returned.
+"""
+function find_multipart_boundaries(bytes::AbstractVector{UInt8}, boundary::AbstractVector{UInt8}; start=1)
+    idxs = []
+    while true
+        (is_terminating_delimiter, i, end_index) = find_multipart_boundary(bytes, boundary; start = start)
+        push!(idxs, (i, end_index))
+        is_terminating_delimiter && break
+        start = end_index + 1
+    end
+    return idxs
+end
+
+"""
+    find_header_boundary(bytes)
+
+Find the end of the multipart header in the byte array.  Returns a Tuple with the
+start index(1) and the end index. Headers are separated from the body by CRLFCRLF.
+
+[RFC2046 5.1](https://tools.ietf.org/html/rfc2046#section-5.1)
+[RFC822 3.1](https://tools.ietf.org/html/rfc822#section-3.1)
+"""
+function find_header_boundary(bytes::AbstractVector{UInt8})
```

Comment:
I feels like this "find subsequence" operation should be in `Base`, but I can't find it.

The `Base._search` function seems to do what you need but it's internal.

Reply:
I agree and there is a "findnext" method which sounds promising, but it operates on an array of Bools and finds the next true index.

I hadn't looked at this code closely before, I noticed that it's using `bytes[i:end_index]` which appears to create a new vector.  Being in a loop that might cause a surge in garbage collection?  I'm not sure if I should change it to something like `view(bytes, i:end_index)`.

Reply:
I noticed this too but I thought I would avoid nitpicking at that :-)

If you do want to make this allocation free, the most straightforward way will just be to hard code the four values which are searched for. (I believe even `@view`/`view` do allocate a small amount for compiler-technical reasons, though this may be fixed in julia 1.4)